### PR TITLE
Background Work

### PR DESCRIPTION
**Pull request by buh**
1. Fixed the background workflow.
2. Fixed the local notifications warning.

### CODE REVIEW

File: Sources/Client/WebSocket/WebSocket.swift [hunk 1]
```diff
@@ -172,30 +167,30 @@ extension WebSocket {
         }
     }
     
-    func disconnect() {
+    func disconnect(reason: String) {
```
Comment (b-onc) [inline Sources/Client/WebSocket/WebSocket.swift]: I'd say a `WebSocketDisconnectReason` enum is needed here

File: Sources/Client/Utils/RepeatingTimer.swift [hunk 2]
```diff
@@ -16,19 +16,19 @@ final class RepeatingTimer {
     
     typealias EventHandler = () -> Void
     
-    private let queue: DispatchQueue?
+    private let queue: DispatchQueue
     private var state: State = .suspended
     let timeInterval: DispatchTimeInterval
-    var eventHandler: EventHandler?
+    var eventHandler: EventHandler
```
Comment (b-onc) [inline Sources/Client/Utils/RepeatingTimer.swift]: Why don't we keep eventHandler as optional if we're going to set it to `{}`?

File: Sources/Client/WebSocket/WebSocket.swift (hunk 1 above)
Comment (buh) [reply Sources/Client/WebSocket/WebSocket.swift]: I didn't think about that when I worked on a fix. I had to find a way to see the flow with disconnection from different points and they all different, so it won't be needed to reuse (like you suggest use enum). I like the current solution with the freedom to set a const string directly in the interesting points.

File: Sources/Client/Utils/RepeatingTimer.swift (hunk 2 above)
Comment (buh) [reply Sources/Client/Utils/RepeatingTimer.swift]: I updated. Please refresh.

File: Sources/Client/Client/Client+Setup.swift [hunk 3]
```diff
@@ -132,7 +132,7 @@ extension Client {
         self.isExpiredTokenInProgress = isExpiredTokenInProgress
         
         if webSocket.isConnected {
-            webSocket.disconnect()
+            webSocket.disconnect(reason: "touched token provider")
```
Comment (b-onc) [inline Sources/Client/Client/Client+Setup.swift]: "Requesting new token..."

File: Sources/Client/Client/Client+Setup.swift [hunk 4]
```diff
@@ -141,7 +141,7 @@ extension Client {
     
     private func setup(token: Token, _ completion: Client.OnConnect?) {
         if webSocket.isConnected {
-            webSocket.disconnect()
+            webSocket.disconnect(reason: "setup token")
```
Comment (b-onc) [inline Sources/Client/Client/Client+Setup.swift]: Setting up new token...

File: Sources/Client/Client/Client.swift [hunk 5]
```diff
@@ -224,8 +223,7 @@ public final class Client {
     
     /// Disconnect the websocket and reset states.
     func reset() {
-        logger?.log("Disconnecting...")
-        webSocket.disconnect()
+        webSocket.disconnect(reason: "reset")
```
Comment (b-onc) [inline Sources/Client/Client/Client.swift]: Resetting connection...

File: Sources/Client/WebSocket/WebSocket.swift [hunk 6]
```diff
@@ -72,7 +70,7 @@ final class WebSocket {
     deinit {
         if isConnected {
             logger?.log("💔 Disconnect on deinit")
-            disconnect()
+            disconnect(reason: "on deinit")
```
Comment (b-onc) [inline Sources/Client/WebSocket/WebSocket.swift]: Deallocating WebSocket...

File: Sources/Client/WebSocket/WebSocket.swift [hunk 7]
```diff
@@ -134,36 +134,31 @@ extension WebSocket {
     }
     
     func disconnectInBackground() {
+        webSocket.callbackQueue.async(execute: disconnectInBackgroundInWebSocketQueue)
+    }
+    
+    private func disconnectInBackgroundInWebSocketQueue() {
         guard stayConnectedInBackground else {
-            disconnect()
+            disconnect(reason: "in background: stayConnectedInBackground is disabled")
```
Comment (b-onc) [inline Sources/Client/WebSocket/WebSocket.swift]: Going into background, stayConnectedInBackground is disabled

File: Sources/Client/WebSocket/WebSocket.swift [hunk 8]
```diff
@@ -134,36 +134,31 @@ extension WebSocket {
...
         
-        guard backgroundTask == .invalid else {
-            return
+        if backgroundTask != .invalid {
+            UIApplication.shared.endBackgroundTask(backgroundTask)
         }
         
         backgroundTask = UIApplication.shared.beginBackgroundTask { [weak self] in
-            self?.disconnect()
+            self?.disconnect(reason: "finished in background")
```
Comment (b-onc) [inline Sources/Client/WebSocket/WebSocket.swift]: Processing finished in background

File: Sources/Client/WebSocket/WebSocket.swift [hunk 9]
```diff
@@ -134,36 +134,31 @@ extension WebSocket {
...
-                self?.disconnect()
-            }
-            
-            webSocket.callbackQueue.asyncAfter(deadline: .now() + WebSocket.maxBackgroundTime, execute: goingToDisconnect)
-            self.goingToDisconnect = goingToDisconnect
-            logger?.log("💜 Background mode on")
-        } else {
-            disconnect()
+        if backgroundTask == .invalid {
+            disconnect(reason: "in background: can't make a background task")
```
Comment (b-onc) [inline Sources/Client/WebSocket/WebSocket.swift]: Can't create a background task

File: Sources/Client/WebSocket/WebSocket.swift [hunk 10]
```diff
@@ -134,36 +134,31 @@ extension WebSocket {
...
         }
     }
     
-    func cancelBackgroundWork() {
-        goingToDisconnect?.cancel()
-        goingToDisconnect = nil
+    private func cancelBackgroundWork() {
+        logger?.log("Cancel background work")
```
Comment (b-onc) [inline Sources/Client/WebSocket/WebSocket.swift]: Cancelling background work

File: Sources/Client/WebSocket/WebSocket.swift [hunk 11]
```diff
@@ -172,30 +167,30 @@ extension WebSocket {
...
             webSocket.disconnect(forceTimeout: 0)
-            logger?.log("Disconnecting...")
         } else {
             logger?.log("Skip disconnecting: WebSocket was not connected")
             connection = .disconnected(nil)
         }
     }
     
     private func clearStateAfterDisconnect() {
+        logger?.log("Clear state after disconnect")
```
Comment (b-onc) [inline Sources/Client/WebSocket/WebSocket.swift]: Clearing state after disconnect

**Pull request merged**